<?xml version="1.0"?>

<launch>


	<!-- =========================================================================================================== -->
	<!-- =========================================================================================================== -->
	<!-- Publish the calibration -->
	<node name="$(arg front_lens_name)" ns="$(arg camera_name)" pkg="kalibr_extended_camera_info_publisher"
			type="camera_info_publisher_node" >
		<param name="camera_ns" value="$(arg front_lens_name)"/>
		<param name="frame_id" value="$(arg camera_name)_$(arg front_lens_name)_optical_frame"/>
		<rosparam command="load" file="$(find camera360_perception)/cfg/cameras_info/insta360_air/$(arg front_lens_name).yaml" subst_value="true" />
	</node>

	<node name="$(arg rear_lens_name)" ns="$(arg camera_name)" pkg="kalibr_extended_camera_info_publisher"
			type="camera_info_publisher_node">
		<param name="camera_ns" value="$(arg rear_lens_name)"/>
		<param name="frame_id" value="$(arg camera_name)_$(arg rear_lens_name)_optical_frame"/>
		<rosparam command="load" file="$(find camera360_perception)/cfg/cameras_info/insta360_air/$(arg rear_lens_name).yaml" subst_value="true" />
	</node>
	<!-- =========================================================================================================== -->



	<node pkg="nodelet" type="nodelet" name="usb_cam"
		args="load usb_cam/UsbCamNodelet $(arg camera_name)_nodelet_manager" output="screen" >
		<param name="video_device" value="/dev/video0" />
		<param name="image_width" value="3008" />
		<param name="image_height" value="1504" />
		<param name="pixel_format" value="mjpeg" />
		<param name="io_method" value="mmap"/>
		<param name="camera_frame_id" value="usb_cam" />
		<remap from="/usb_cam/image_raw" to="/$(arg camera_name)/image_raw"/>
	</node>
	<node pkg="rosservice" type="rosservice" name="global_loc" args="call --wait /usb_cam/toggle_capture true" />


	<arg name="split_image_width" default="1504" />
	<arg name="split_image_height" default="1504" />

	<node pkg="nodelet" type="nodelet" name="$(arg camera_name)_split_front" 
			args="load image_proc/crop_decimate $(arg camera_name)_nodelet_manager" output="screen" respawn="true">

		<param name="queue_size" type="int" value="1" />
		<param name="x_offset" type="int" value="0" />
		<param name="y_offset" type="int" value="0" />
		<param name="width" type="int" value="$(arg split_image_width)" />
		<param name="height" type="int" value="$(arg split_image_height)" />
		<!-- <param name="camera_name" value="front"/> -->
		<param name="target_frame_id" value="$(arg camera_name)_$(arg front_lens_name)_optical_frame"/>

		<remap from="camera/image_raw" to="/$(arg camera_name)/image_raw"/>
		<remap from="camera/image_info" to="/$(arg camera_name)/camera_info"/>

		<!-- Remap output to front image -->
		<remap from="camera_out/image_raw" to="/$(arg camera_name)/$(arg front_lens_name)/image_raw" />
		<!-- Dont use original camera info -->
		<remap from="/stereo/right/camera_info" to="/null/right/camera_info" />
	</node>

	<!-- Extract rear image -->
	<node pkg="nodelet" type="nodelet" name="$(arg camera_name)_split_rear" 
			args="load image_proc/crop_decimate $(arg camera_name)_nodelet_manager" output="screen" respawn="true">

		<!-- <param name="camera_info_url" value="$(arg camera_info_url_left)" /> -->
		<param name="queue_size" type="int" value="1" />
		<param name="x_offset" type="int" value="$(arg split_image_width)" />
		<param name="y_offset" type="int" value="0" />
		<param name="width" type="int" value="$(arg split_image_width)" />
		<param name="height" type="int" value="$(arg split_image_height)" />
		<!-- <param name="camera_name" value="rear"/> -->
		<param name="target_frame_id" value="$(arg camera_name)_$(arg rear_lens_name)_optical_frame"/>

		<remap from="camera/image_raw" to="/$(arg camera_name)/image_raw"/>
		<remap from="camera/image_info" to="/$(arg camera_name)/camera_info"/>

		<!-- Remap output to rear image -->
		<remap from="camera_out/image_raw" to="/$(arg camera_name)/$(arg rear_lens_name)/image_raw" />
		<!-- Dont use original camera info -->
		<remap from="/stereo/left/camera_info" to="/null/left/camera_info" />
	</node>
</launch>